feat(discover/perf/alerts) Add OR / AND / () support to query filters

evanh opened the pull request:
Users can now add the 'OR' and 'AND' keyword to their query filters, as well as group their filters using '()'.  This functionality will only be available for Discover, Performance and Metric Alerts.

evanh committed:
feat(discover/perf/alerts) Add OR / AND / () support to query filters

Users can now add the 'OR' and 'AND' keyword to their query filters, as well as
group their filters using '()'.  This functionality will only be available for
Discover, Performance and Metric Alerts.

silent1mezzo commented on tests/sentry/api/test_event_search.py:
```diff
@@ -768,271 +768,801 @@ def test_empty_string(self):
-            SearchBoolean(left_term=self.term1, operator="OR", right_term=self.term2)
-        ]
+        result = get_filter("user.email:foo@example.com OR user.email:bar@example.com")
+        print("RES", result.conditions)
```
Extra print statement

silent1mezzo commented on tests/sentry/api/test_event_search.py:
```diff
@@ -768,271 +768,801 @@ def test_empty_string(self):
-                ),
-            )
+        )
+        print("RESULT", result.conditions)
```
Extra print statement

silent1mezzo commented on tests/sentry/api/test_event_search.py:
```diff
@@ -768,271 +768,801 @@ def test_empty_string(self):
-        assert parse_search_query("user.email:foo@example.com OR user.email:bar@example.com") == [
-            SearchBoolean(left_term=self.term1, operator="OR", right_term=self.term2)
-        ]
+        result = get_filter("user.email:foo@example.com OR user.email:bar@example.com")
```
More of a question, is there any reason why we don't turn these emails into variables and use them throughout tests? I doubt they'd ever change but if they did for some reason it'd be a lot of broken tests.

It'd also have a benefit of making the asserts easier to read since you don't need to store which email refers to which term in your head.

evanh replied:
I've been thinking about the best way to do this and I may revisit it. I was worried about hiding too much of the details in order to try and limit the "noise" in a test, but I'll look these over again and see if it can be clearer what things are so the expected result is more obvious.

markstory replied:
Would renaming `term1` etc. to `foo_term`, `bar_term` help? Then the names of the email address and term attribute would have something in common.

silent1mezzo commented on tests/sentry/api/test_event_search.py:
```diff
@@ -768,271 +768,801 @@ def test_empty_string(self):
+        )
+        print("RESULT", result.conditions)
+        assert result.conditions == [
+            [["or", [self.term1, ["and", [self.term2, self.term3]]]], "=", 1]
```
As an example to turning the emails into variables, I assume this is correct but without looking above I'm not 100% sure that `foobar@example.com` is term1.

evanh replied:
Maybe the solution is to rename these to `self.foo`, `self.bar` etc to make it more obvious what the expected case is.

markstory replied:
I agree that would help.

silent1mezzo commented on tests/sentry/api/test_event_search.py:
```diff
@@ -768,271 +768,801 @@ def test_empty_string(self):
+            get_filter("(count():>1 AND a:b) OR a:b")
+        assert (
+            six.text_type(error.value)
+            == "Having an OR between aggregate filters and normal filters is invalid."
```
If my understanding is correct it's the `(count():>1 AND a:b)` that isn't allowed.

```suggestion
            == "Having an AND between aggregate filters and normal filters is invalid."
```

evanh replied:
No it's the OR that is not allowed. Aggregate filters (`count():>1`) are put in the HAVING clause of the Clickhouse query while other filters are put in the WHERE clause. Those are implicitly ANDed together, but you can't OR between the two. Thus doing an OR between an aggregate filter and normal filter isn't allowed.

evanh commented on src/sentry/api/event_search.py:
```diff
@@ -831,6 +823,223 @@ def convert_search_filter_to_snuba_query(search_filter, key=None):
+        converted_filter = convert_aggregate_filter_to_snuba_query(term, params)
+        return None, convert_condition_to_function(converted_filter), projects_to_filter, group_ids
+
+    return None, None, projects_to_filter, group_ids
```
Right now it only gets directly unpacked into local variables. I don't know that there are any list index accesses. I do think adding some namedtuples or something around these values would be very useful, but I think that work is out of scope for this.

fpacifici commented on src/sentry/api/event_search.py:
```diff
@@ -107,15 +106,15 @@ def translate(pat):
 
 event_search_grammar = Grammar(
```
Does this apply to issue search as well ? 
That would be a problem because, depending on the filters applied the query may be served entirely by snuba (which would be fine) or by postgres + snuba by doing the join in python.
Queries like 
`assigned_to me OR assigned_to you AND event.message = 'asd'`
won't work because they would turn into 
`(assigned_to me OR assigned_to you) AND event.message = 'asd'`

evanh replied:
This does not apply to issue search. Using OR/AND in issue search will result in an error.

markstory commented on src/sentry/api/event_search.py:
```diff
@@ -860,97 +1077,41 @@ def get_filter(query=None, params=None):
+        for term in parsed_terms
+    ):
+        # TODO evanh: We can remove all top level ANDs and extend the conditions/having to
+        # avoid unnecesary nesting, e.g. [["and", [["and", [a, b]], ["and", [c, d]]]]] -> [a, b, c, d]
```
This seems like a nice optimization to do once this change lands.

markstory commented on tests/sentry/api/test_event_search.py:
```diff
@@ -768,271 +768,917 @@ def test_empty_string(self):
+            get_filter("a:b AND c:d AND")
+
+        with self.assertRaisesRegexp(
+            InvalidSearchQuery, "Condition is missing on the left side of 'OR' operator"
```
Would be super :rainbow: if we could tell them the column offset of the problem as well in the future.

markstory reviewed (approved):
Looks good to me. We should hold of on updating the documentation until we have the frontend query manipulations figured out as the current manipulation tools we have won't handle OR or parenthetical expressions.

evanh committed:
Merge branch 'master' into evanh/feat/boolean-paren-in-discover-search

evanh committed:
Merge branch 'master' into evanh/feat/boolean-paren-in-discover-search

evanh committed:
Merge branch 'master' into evanh/feat/boolean-paren-in-discover-search

evanh committed:
rename some terms to make tests more readable

evanh commented:
> Will we have a way to track errors generated by boolean search? Would be good to save the search conditions if so.

Not specifically, but we record errors in general, which will have the conditions in them.

evanh committed:
Merge branch 'master' into evanh/feat/boolean-paren-in-discover-search

evanh committed:
Merge branch 'master' into evanh/feat/boolean-paren-in-discover-search

evanh committed:
Merge branch 'master' into evanh/feat/boolean-paren-in-discover-search

evanh committed:
remove xrange

evanh committed:
Merge branch 'master' into evanh/feat/boolean-paren-in-discover-search

markstory commented on tests/sentry/api/test_event_search.py:
```diff
@@ -768,271 +768,902 @@ def test_empty_string(self):
+        self.bar = ["equals", ["user.email", "bar@example.com"]]
+        self.foobar = ["equals", ["user.email", "foobar@example.com"]]
+        self.hello = ["equals", ["user.email", "hello@example.com"]]
+        self.hi = ["equals", ["user.email", "hi@example.com"]]
```
This helps make the tests much easier to understand. :clap:

evanh committed:
Merge branch 'master' into evanh/feat/boolean-paren-in-discover-search

evanh committed:
feat(search) Change tokenizeSearch to allow for OR/AND/() (#19854)

Rewrite tokenizeSearch so that it can tokenize and stringify OR/AND and grouped
expressions correctly. Also fix a bug where all raw string searches would be
grouped together, which would change the query being run if the user wanted
to search for two separate strings.

E.g. `a b c:c d` should convert to `message:"a b" AND c:c AND message:"d"`.
Since tokenizeSearch was grouping the text components, the query would get
transformed to:
`a b d c:c` which converts to `message:"a b d" AND c:c`

evanh committed:
Merge branch 'master' into evanh/feat/boolean-paren-in-discover-search
